### PR TITLE
Integrate revisions with event dispatcher

### PR DESCRIPTION
**Pull request by JoelAlphonso**
Merging `Joel_feat_revision-event` into `joel_feat_event-dispatcher` (#64).

### Features

* **object:** add an ObjectServiceProvider for object package services ([117dad5](https://github.com/charcoalphp/charcoal/commit/117dad5a80b32bc3a18dece45f39e4f05a05dda8))
* **revision:** add `properties` and `propertyBlacklist` options for Revisions ([f2cf7c9](https://github.com/charcoalphp/charcoal/commit/f2cf7c9d80c0c30490310ae0a7060153978e73e3))
* **revision:** add a listener to hook revision generation ([4cb6fbb](https://github.com/charcoalphp/charcoal/commit/4cb6fbb1ce10f0f85c8686419f02124d53473163))
* **revision:** add a revision service to handle all revision related operations ([0936d20](https://github.com/charcoalphp/charcoal/commit/0936d20e320aea239972bd6e5d312c6486ef563c))
* **revision:** add RevisionConfig.php to cast revisions config nodes to ([2bc3ece](https://github.com/charcoalphp/charcoal/commit/2bc3eced6a52b6ca17b9fd7419baeb1a831e9481))
* **revision:** remove revision generation from Content.php ([8fb6b25](https://github.com/charcoalphp/charcoal/commit/8fb6b257858b16b0c17727691fc5af920488f2f3))
* **revision:** use the revisions service everywhere and refactor it to require setting a model on the service ([aef32f0](https://github.com/charcoalphp/charcoal/commit/aef32f05c28e6a6b09f4df09d61bddb1d12e05f3))

### Bug Fixes

* **revision:** fix included properties and properties options ([feddd68](https://github.com/charcoalphp/charcoal/commit/feddd68103900fd2ecd701fc540b523d47cf8c98))

### CODE REVIEW

File: packages/object/src/Charcoal/Object/GenerateRevisionListener.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+use Pimple\Container;
+
+/**
+ * Listener
+ */
+class GenerateRevisionListener extends AbstractEventListener
+{
+    protected RevisionService $revisionService;
+
+    public function __invoke(object $event)
```
Comment (mcaskill) [inline packages/object/src/Charcoal/Object/GenerateRevisionListener.php:16]: I think this should be type-hinted to `Charcoal\Event\Events\Object\AbstractObjectEvent` which, at this point, should also be an interface (`ObjectEventInterface` that declares `getObject()`).

```suggestion
    public function __invoke(object $event)
```

File: packages/object/src/Charcoal/Object/ObjectServiceProvider.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+                $configData = [
+                    'enabled' => $configData,
+                ];
+            }
+
+            return new RevisionConfig($configData);
+        };
+
+        $pimple['revision/service'] = function (Container $pimple): RevisionService {
+            $services = new ServiceLocator($pimple, ['revision/config', 'model/factory', 'logger']);
```
Comment (mcaskill) [inline packages/object/src/Charcoal/Object/ObjectServiceProvider.php:35]: Good idea. Its about time we start using PSR-11 😆

Should this be introduced into the service provider (and factories?) for [charcoal/event](#64)?

Also, for future diffs:

```suggestion
            $services = new ServiceLocator($pimple, [
                'revision/config',
                'model/factory',
                'logger',
            ]);
```

**Review by mcaskill — CHANGES_REQUESTED** [review]
Should the class names use `Revision` or `Revisions` plural?

- `RevisionsServiceProvider`
- `RevisionsConfig`
- `RevisionsModelConfig`
- `RevisionsManager`

File: packages/object/src/Charcoal/Object/ObjectServiceProvider.php (hunk 2 above)
Comment (JoelAlphonso) [reply packages/object/src/Charcoal/Object/ObjectServiceProvider.php:35]: 👍

File: packages/object/src/Charcoal/Object/RevisionModelConfig.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+     * @return array
+     */
+    public function getIncludedProperties(): array
+    {
+        return $this->includedProperties;
+    }
+
+    public function hasIncludedProperties(): bool
+    {
+        return !!count($this->includedProperties);
```
[the remark this answers is not in the document]
Comment (JoelAlphonso) [reply packages/object/src/Charcoal/Object/RevisionModelConfig.php:73]: Yup, je suis d'accord

File: packages/object/src/Charcoal/Object/RevisionService.php [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+ * Revision Service
+ *
+ * Service handling revision generation and retrieval.
+ * Can be implemented in a listener
+ *
+ * Revisions need to act on a ModelInterface,
+ * So to use the revision service, one have to set a ModelInterface beforehand.
+ * Failure to do so will result in a
+ */
+class RevisionService
```
[the remark this answers is not in the document]
Comment (JoelAlphonso) [reply packages/object/src/Charcoal/Object/RevisionService.php:23]: Done

File: packages/object/src/Charcoal/Object/GenerateRevisionListener.php (hunk 1 above)
Comment (JoelAlphonso) [reply packages/object/src/Charcoal/Object/GenerateRevisionListener.php:16]: I like the interface thing. I'll refactor object events in the event branch

File: packages/object/src/Charcoal/Object/RevisionModelConfig.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public function hasExcludedProperties(): bool
+    {
+        return !!count($this->excludedProperties);
+    }
+
+    /**
+     * @return array
+     */
```
[the remark this answers is not in the document]
Comment (JoelAlphonso) [reply packages/object/src/Charcoal/Object/RevisionModelConfig.php:65]: Haaa oups, the comments were autogenerated, my bad

**Pull request merged**